url8af84d296f_mrcosta_gantt_generator_tgz-pJ8-org_gantt_generator_mrcpsp_RunJ8/src/org/swiftgantt/model/Task.java

```diff
@@ -0,0 +1,622 @@
+			ret = progress;
+		} else if (timeUnit == TimeUnit.Year) {
+			ret = progress;
+		}
+		return ret;
+	}
+
+	/**
+	 * Get duration of this <code>Task</code>, exclude the rest-out time.
+	 * 
+	 * @return the duration
+	 */
+	public int getDuration() {
+		// The duration that the task cost, it will be determined by the TimeUnit, Start time and End time.
+		// Consider to change this implementaion, because it will calculate the duration every time.
+		return this.calcDuration();
+	}
+
+	/**
+	 * Get progress of this <code>Task</code>. If progress exceeds the task duration, it will be limited within the duration.
+	 * 
+	 * @return the progress
+	 */
+	public int getProgress() {
+		if (progress < 0) {
+			return 0;
+		}
+		int duration = this.getDuration();
+		if (progress > duration) {
+			return duration;
+		}
+		return progress;
+	}
+
+	/**
+	 * Set progress of this <code>Task</code>.
+	 * 
+	 * @param progress the progress to set
+	 */
+	public void setProgress(int progress) {
+		Object oldValue = this.progress;
+		this.progress = progress;
+		//    	logger.debug(" Progress for " + this.name + " " + this.progress + " by " + this.duration);
+		this.firePropertyChanged("Progress", oldValue, progress);
+	}
+
+	/**
+	 * Get start time of this <code>Task</code>.
+	 * 
+	 * @return the start
+	 */
+	public Time getStart() {
+		return this.checkAndLimitStartTime(this.start);
+	}
+
+	/**
+	 * Set start time of this <code>Task</code>.
+	 * @param start
+	 */
+	public void setStart(Calendar start){
+		this.setStart(new Time(start));
+	}
+
+	/**
+	 * Set start time of this <code>Task</code>.
+	 * 
+	 * @param start the start to set
+	 * @since 0.4.0
+	 */
+	public void setStart(Time start) {
+		if (start == null) {
+			return;
+		}
+		Object oldValue = this.start;
+		this.start = start;// Clone to not to change the start time for outside.
+		if (this.end == null) {
+			this.end = start;
+		}
+		// Adjust actual start time by predecessor task
+		if (this.predecessors.size() > 0) {
+			Task p = super.getLatestPredecessor();
+			if (p != null && (p.getEnd().after(start) || p.getEnd().equals(start))) {
+				adjustActualTimesByPredecessor(p);
+			} else {
+				this.actualStart = start.clone();
+			}
+		} else {
+			this.actualStart = start.clone();
+		}
+		this.firePropertyChanged("Start Time", oldValue, start);
+	}
+
+	/**
+	 * Handler the predecessor changes event.
+	 */
+	public void predecessorChanged(PredecessorChangeEvent e) {
+		Task p = (Task) e.getPredecessor();
+		if (p != null && p.getEnd() != null && p.getEnd().after(this.getStart())) {
+			this.adjustActualTimesByPredecessor();
+		}
+	}
+
+	/**
+	 * Adjust actual start and end time by it's latest predecessor.
+	 */
+	public void adjustActualTimesByPredecessor() {
+		Task p = this.getLatestPredecessor();
+		this.adjustActualTimesByPredecessor(p);
+	}
+
+	/*
+	 * Adjust actual start and end time by specified predecessor;
+	 */
+	protected void adjustActualTimesByPredecessor(Task predecessor) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.